### PR TITLE
fix: run blocking call in threadpool

### PR DESCRIPTION
**Pull request by kumaraditya303**

### CODE REVIEW

File: playwright/_impl/_browser_context.py [hunk 1]
```diff
@@ -298,8 +299,7 @@ async def _pause(self) -> None:
     async def storage_state(self, path: Union[str, Path] = None) -> StorageState:
         result = await self._channel.send_return_as_dict("storageState")
         if path:
-            with open(path, "w") as f:
-                json.dump(result, f)
+            await async_writefile(path, "w", json.dumps(result))
```
Comment (pavelfeldman) [inline playwright/_impl/_browser_context.py:302]: I was already leaving this comment, should be `async_write_file(page, json.dumps(result))`

File: playwright/_impl/_browser.py [hunk 2]
```diff
@@ -214,5 +215,4 @@ def normalize_context_params(is_sync: bool, params: Dict) -> None:
     if "storageState" in params:
         storageState = params["storageState"]
         if not isinstance(storageState, dict):
-            with open(storageState, "r") as f:
-                params["storageState"] = json.load(f)
+            params["storageState"] = json.loads(await async_read_text(storageState))
```
Comment (pavelfeldman) [inline playwright/_impl/_browser.py:218]: `(await async_read_file()).decode()`

File: playwright/_impl/_element_handle.py [hunk 3]
```diff
@@ -241,8 +242,7 @@ async def screenshot(
         decoded_binary = base64.b64decode(encoded_binary)
         if path:
             make_dirs_for_file(path)
-            with open(path, "wb") as fd:
-                fd.write(decoded_binary)
+            await async_writefile(path, "wb", decoded_binary)
```
Comment (pavelfeldman) [inline playwright/_impl/_element_handle.py:245]: ditto

File: playwright/_impl/_file_chooser.py [hunk 4]
```diff
@@ -57,20 +58,19 @@ async def set_files(
...
-                    {
-                        "name": os.path.basename(item),
-                        "buffer": base64.b64encode(fd.read()).decode(),
-                    }
-                )
+        if isinstance(item, (str, Path)):
+            file_payloads.append(
+                {
+                    "name": os.path.basename(item),
+                    "buffer": base64.b64encode(await async_read_binary(item)).decode(),
```
Comment (pavelfeldman) [inline playwright/_impl/_file_chooser.py:71]: `async_read_file`

File: playwright/_impl/_helper.py [hunk 5]
```diff
@@ -231,3 +231,27 @@ def make_dirs_for_file(path: Union[Path, str]) -> None:
...
+) -> None:
+    def inner() -> None:
+        with open(file, mode) as fh:
+            fh.write(data)
+
+    loop = asyncio.get_running_loop()
+    await loop.run_in_executor(None, inner)
+
+
+async def async_read_binary(file: Union[str, Path]) -> bytes:
```
Comment (pavelfeldman) [inline playwright/_impl/_helper.py:247]: `async_read_file`

File: playwright/_impl/_helper.py [hunk 6]
```diff
@@ -231,3 +231,27 @@ def make_dirs_for_file(path: Union[Path, str]) -> None:
...
+    def inner() -> bytes:
+        with open(file, "rb") as fh:
+            return fh.read()
+
+    loop = asyncio.get_running_loop()
+    return await loop.run_in_executor(None, inner)
+
+
+async def async_read_text(file: Union[str, Path]) -> str:
+    return (await async_read_binary(file)).decode()
```
Comment (pavelfeldman) [inline playwright/_impl/_helper.py:257]: inline this, no reason to have a separate function.

File: playwright/_impl/_browser_context.py (hunk 1 above)
Comment (kumaraditya303) [reply playwright/_impl/_browser_context.py:302]: Okay but how would you write in binary with this ?

File: playwright/_impl/_helper.py (hunk 6 above)
Comment (kumaraditya303) [reply playwright/_impl/_helper.py:257]: Sure!

File: playwright/_impl/_browser_context.py (hunk 1 above)
Comment (kumaraditya303) [reply playwright/_impl/_browser_context.py:302]: OR you mean we should have two functions for writing one for str and one for bytes ?

Comment (pavelfeldman) [reply playwright/_impl/_browser_context.py:302]: `async_write_file` would accept `Union[str, bytes]` and will pick the right mode.

Comment (kumaraditya303) [reply playwright/_impl/_browser_context.py:302]: Sure! Thanks for confirming

**Review by pavelfeldman — APPROVED** [review]

**Pull request merged**